DM-5466: obs_decam calibration ingest uses fragile relative paths

mrawls opened the pull request.

mrawls commented:
Thanks for pointing this out @PaulPrice. I was under the impression that the main issue here is the registry not containing full (absolute) filepaths. Is the preferred behavior for the calib registry to always contain paths relative to the calib directory?

PaulPrice commented:
All paths in the butler are relative to the data repository, except that calib paths are relative to the calib repository.  Of course, most datasets don't have `%(path)s` in the template.  The exception is the "community pipeline"-produced calibs.  I don't know what the status is of those (@hsinfang would know), but I personally recommend building your own calibs using the LSST pipeline.  In any case, I don't think the CP-produced calibs should be an exception to the rule of paths relative to the calib repo.

hsinfang commented:
Using CP-produced calibs should work just fine, or at least it did last time I manually tried >1 month ago. It's not CI-ed afaik. 

I second @ctslater that having paths defined relative to the calib root is already an improvement over the current situation. 

Making symlinks to CP files during ingestion and using templates more properly without `%(path)s` sounds even better (I think that's what @ctslater suggested on [Jira](https://jira.lsstcorp.org/browse/DM-5466?focusedCommentId=68846&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-68846) ).  Note that [obs_cfht](https://github.com/lsst/obs_cfht/blob/master/policy/MegacamMapper.paf#L75) also uses `%(path)s` currently. I hope the solution can be applied to both obs_decam and obs_cfht.

mrawls commented:
Thanks for this discussion. My use case is presently ingesting CP calibs. After working through bits of this with @ctslater and @SimonKrughoff on Friday, I think these latest commits should address all the issues.

ctslater commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -105,3 +82,73 @@ def getExtensionName(md):
+    clobber = Field(dtype=bool, default=False, doc="Clobber existing file?")
+
+
+class DecamIngestCalibsTask(IngestCalibsTask):
```
Can these changes get folded back into pipe_tasks's IngestCalibsTask? It would be nice not to bifurcate the process per camera if we don't have to.

mrawls replied:
I'm happy to do that, but since it would change the behavior of how all calibrations are ingested (symlinking the calib files) beyond just obs_decam, is this the kind of thing that should go through an RFC first?

hsinfang replied:
I would prefer this going into pipe_tasks's IngestCalibsTask too.  

I think at this moment the only other customer of pipe_tasks's IngestCalibsTask is obs_subaru, but yeah we could be more certain an RFC.

ctslater replied:
Yes it would require an RFC, but I think it would be an improvement for the existing script since it would better parallel the behavior of non-calib ingest. The current behavior is equivalent to mode=skip.

hsinfang replied:
I'm not sure if an RFC is really needed to augment features in this case though. Looks like it could be all backward compatible.

ctslater replied:
Ah, I guess it could be backwards compatible if the default was set to mode=skip, but then that default wouldn't work properly on decam (it would register the files, but since it doesn't store the paths anymore, they wouldn't be findable later). I would be hesitant for a user-facing behavior like this have different per-camera defaults.

Does obs_cfht use ingestCalibs.py? They have paths in their calib templates, but I can't see how they ingest those.

hsinfang replied:
Looks to me that `obs_cfht` uses its own `bin/megacamCalibRegistry.py` to create `calibRegistry.sqlite3`.  Also obs_cfht calib templates predates `IngestCalibsTask`.

mrawls replied:
OK, I think it makes sense to put it in pipe_tasks and set the default mode=skip for backwards compatibility. What is your concern about this option not working properly with decam, @ctslater? I can edit the obs_decam README (should probably do that anyway!) and clarify that setting mode=link is necessary/recommended. But if you all prefer to have the default behavior be mode=link then I'm happy to file an RFC and do it that way.

ctslater replied:
mode=skip would "work", but it wouldn't do anything particularly useful for decam. The butler still wouldn't know how to find the calibs. I'm just advocating for setting useful defaults, particularly if we can't find a use case for the current default.

ctslater commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -105,3 +82,73 @@ def getExtensionName(md):
         """
         return md.get('EXTNAME')
+
+#     def getDestination(self, butler, info, filename):
```
`getDestination()` can stay here. If you look at obs_decam/config/ingestCalibs.py, the first two line are retargeting the IngestCalibTask's parse task to this decam-specific one. This function will effectively replace the simplified version in pipe_tasks (which I left a comment about over there). This file will then only contain a `DecamCalibsParseTask`, and I don't think any other classes.

hsinfang commented on config/ingestCalibs.py:
```diff
@@ -1,6 +1,8 @@
 from lsst.obs.decam.ingestCalibs import DecamCalibsParseTask
 config.parse.retarget(DecamCalibsParseTask)
 config.parse.hdu = 1
+config.parse.translation = {"obstype": "OBSTYPE"}
```
Where is this used? Isn't `OBSTYPE` obtained directly in `getCalibType`? If I understand correctly, this line should not be needed.

hsinfang commented on config/ingestCalibs.py:
```diff
@@ -14,7 +16,6 @@
 config.register.columns = {'filter': 'text',
                            'ccdnum': 'int',
-                           'path': 'text',
                            'calibDate': 'text',
```
If the decision is to continue using `path` in the butler policy template of `defect` or/and `fringe`, I don't think `path` can be removed from the schema.

hsinfang commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -2,39 +2,16 @@
 
 class DecamCalibsParseTask(CalibsParseTask):
-
-    def getInfo(self, filename):
```
I don't think this override can be removed if we want to continue ingesting `defect` the same way before this ticket.  Would it hinder ingesting `bias`/`flat` in the new way?

mrawls replied:
Good catch! It didn't occur to me that this specific getInfo was for handling defects.

hsinfang commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -105,3 +82,26 @@ def getExtensionName(md):
         """
         return md.get('EXTNAME')
+
+    def getDestination(self, butler, info, filename):
```
Would it be possible to use the base method and then add necessary overrides? 

Also I worry what this means to ingest the LSST-generated calibration data for DECam, the scencario @mfisherlevine has been using. Somebody would want to test that scenario before merging.

mrawls replied:
The `getDestination` here is sufficiently different from the one in `pipe_tasks` that I think it makes sense to keep this here explicitly, but please let me know if I'm missing something.

I chatted with Merlin; he did try doing this, but it sounds like his problems occurred before ingestCalibs and it's not (easily) possible to test ingesting LSST-generated calibration data for decam because those calibration products don't exist yet. It's something I'd definitely like to revisit later!

hsinfang commented on README.md:
```diff
@@ -63,6 +63,8 @@ Documentation of the LSST Science Pipelines is at https://pipelines.lsst.io
         $ ingestImagesDecam.py /path/to/repo --filetype raw /path/to/raw/*.fits.fz
         $ ingestCalibs.py /path/to/repo/  --calib /path/to/calib/repo/ --calibType defect /path/to/calib/*fits --validity 999
```
Should the suggestion of using `--mode=link` go here (line 64) directly? Or are `cpBias`/`cpFlat` ingested differently than `defect`/`fringe`?  (Related to some questions below)

mrawls replied:
I will clarify that this suggestion is for ingesting biases and flats only, and not currently supported for fringes and defects. I will also add more examples here for how to ingest each kind of calibration product.

ctslater replied:
I am slightly confused, we still need to be able to ingest fringes and defects. Is there a different method for that?

mrawls replied:
No, the intention is still to use ingestCalibs.py for ingesting biases, flats, fringes, and defects. The suggestion I was referring to above is the `--mode=link` suggestion explicitly; one should just leave `--mode` out of it altogether for fringes and defects. I think the latest commit below (which also includes an update to the README) clarifies this.

mrawls commented:
I think this latest commit addresses your concerns above. I will also update the testdata_decam registries to be consistent with the now-unchanged calib repo database schema.

mrawls commented:
I have now tested this in a few different places to ingest flats, biases, fringes, and defects, and as far as I can tell it works as expected! I rebased to make the history here cleaner and also rebased to master; I think this (and the three other PRs associated with this ticket) should all be ready to merge now.

hsinfang commented on README.md:
```diff
@@ -61,7 +61,15 @@ Documentation of the LSST Science Pipelines is at https://pipelines.lsst.io
 
         $ ingestImagesDecam.py /path/to/repo --filetype raw /path/to/raw/*.fits.fz
-        $ ingestCalibs.py /path/to/repo/  --calib /path/to/calib/repo/ --calibType defect /path/to/calib/*fits --validity 999
+        $ ingestCalibs.py /path/to/repo/  --calib /path/to/calib/repo/ --calibType defect /path/to/calib/*fits --validity 0
```
FYI, the argument `validity` will be ignored for `defect` (or whatever listed in config `validityUntilSuperseded`), so 0 or 999 doesn't matter (DM-5500)

hsinfang commented on python/lsst/obs/decam/ingestCalibs.py:
```diff
@@ -33,8 +36,7 @@ def getInfo(self, filename):
 
     def _translateFromCalibId(self, field, md):
-        """Fetch the ID from the CALIB_ID header
-
```
Doesn't this docstring need to be followed by a blank line?
